hudi-common/src/main/java/org/apache/hudi/common/engine/HoodieEngineContext.java

```diff
@@ -139,15 +139,16 @@ public abstract <I, K, V> List<V> reduceByKey(
   /**
    * Returns reader context factory for write operations in the table.
    *
-   * @param metaClient Table meta client
-   * @param recordType Record type
-   * @param properties Typed properties
+   * @param metaClient            Table meta client
+   * @param recordType            Record type
+   * @param properties            Typed properties
+   * @param outputsCustomPayloads Whether the reader context factory should output custom payloads. Final merging of records before writes does not require custom payloads.
    */
   public ReaderContextFactory<?> getReaderContextFactoryForWrite(HoodieTableMetaClient metaClient, HoodieRecord.HoodieRecordType recordType,
-                                                                 TypedProperties properties) {
+                                                                 TypedProperties properties, boolean outputsCustomPayloads) {
```

Comment:
This flag is only meaningful for avro reader context, is there anyway we can constraint it just to `AvroReaderContextFactory`?

Reply:
we have a plan to abadon the payload based records in writer path right? So this should be just a temporary solution?

Reply:
We'll still need it for ExpressionPayload and for any user provided payload so it is not temporary but these restrictions may allow us to clean things up further.